src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -759,3 +759,248 @@ - (void)invokeCommandWithEndpointID:(NSNumber *)endpointID
+    if (!MTRInputIsStructurallyValid(input, @[ @ { MTRCommandPathKey : MTRCommandPath.class } ])) {
+        MTR_LOG_ERROR("Invoke response is not an array with the right things in it: %@", input);
+        return NO;
+    }
+
+    NSArray * inputArray = input;
+    if (inputArray.count != 1) {
+        MTR_LOG_ERROR("Invoke response is not an array with exactly one entry: %@", input);
+        return NO;
+    }
+
+    // Note that we have already checked the one entry is a dictionary that has
+    // MTRCommandPathKey.
+    NSDictionary * responseValue = inputArray[0];
+    id data = responseValue[MTRDataKey];
+    id error = responseValue[MTRErrorKey];
+
+    if (data != nil && error != nil) {
+        MTR_LOG_ERROR("Invoke response claims to have both data and error: %@", responseValue);
+        return NO;
+    }
+
+    if (error != nil) {
+        return MTRInputIsStructurallyValid(error, NSError.class);
+    }
+
+    if (data == nil) {
+        // This is valid: indicates a success status response.
+        return YES;
+    }
+
+    if (!MTRInputIsStructurallyValid(data, MTRInputStructureDataValue)) {
+        MTR_LOG_ERROR("Invoke response claims to have data that is not a data-value: %@", data);
+        return NO;
+    }
+
+    // Now we know data is a dictionary (in fact a data-value).  The only thing
+    // we promise about it is that it has type MTRStructureValueType.
+    NSDictionary * dataDictionary = data;
+    if (dataDictionary[MTRTypeKey] != MTRStructureValueType) {
+        MTR_LOG_ERROR("Invoke response data is not of structure type: %@", data);
+        return NO;
+    }
+
+    return YES;
+}
+
+BOOL MTRInputIsStructurallyValid(id input, id structure)
+{
+    if ([structure isEqual:MTRInputStructureDataValue]) {
+        return MTRDataValueDictionaryIsWellFormed(input);
+    }
+
+    if ([structure isEqual:MTRInputStructureAttributeReport]) {
+        return MTRAttributeReportIsStructurallyValid(input);
+    }
+
+    if ([structure isEqual:MTRInputStructureEventReport]) {
+        return MTREventReportIsStructurallyValid(input);
+    }
+
+    if ([structure isEqual:MTRInputStructureInvokeResponse]) {
+        return MTRInvokeResponseIsStructurallyValid(input);
+    }
+
+    // Literal dictionaries and arrays actually have classes that are not
+    // exactly NSArray.class and NSDictionary.class, so checking isKindOfClass
+    // against [structure class] needs to be avoided.
+    if ([structure isKindOfClass:NSDictionary.class] &&
+        [input isKindOfClass:NSDictionary.class]) {
+        return MTRDictionaryIsStructurallyValid(input, structure);
+    }
+
+    if ([structure isKindOfClass:NSArray.class] &&
+        [input isKindOfClass:NSArray.class]) {
+        return MTRArrayIsStructurallyValid(input, structure);
+    }
+
+    if ([structure class] != structure) {
+        // Not a class object, not sure what to do with it.
+        MTR_LOG_ERROR("Unknown structure value: %@", structure);
```

Comment:
I don't think we need to tolerate this kind of programmer error, so we could just `MTRThrowInvalidArgument` here

Reply:
This generic method went away.